server.js

```diff
@@ -1,33 +1,235 @@
+  return (req, res, next) => {
+    const {
+      title,
+      author,
+      language,
+      minRating,
+      maxRating,
+      minPages,
+      maxPages,
+    } = req.query;
+
+    const pages = parseInt(req.query.pages);
+    const limit = parseInt(req.query.limit);
+
+    const response = {}; // for sending the response
+
+    let filteredBooksData = booksDataToSend;
+    //* ---------------  filtration based on different search criteria ------*//
+    if (title) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) => book.title.toLowerCase().indexOf(title.toLowerCase()) !== -1
+      );
+    }
+
+    if (author) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) =>
+          book.authors.toLowerCase().indexOf(author.toLowerCase()) !== -1
+      );
+    }
+    if (language) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) =>
+          book.language_code.toLowerCase().indexOf(language.toLowerCase()) !==
+          -1
+      );
+    }
+
+    if (minRating) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) => book.average_rating >= +minRating
+      );
+    }
+
+    if (maxRating) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) => book.average_rating <= +maxRating
+      );
+    }
+    if (minPages) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) => book.num_pages >= +minPages
+      );
+    }
+    if (maxPages) {
+      filteredBooksData = filteredBooksData.filter(
+        (book) => book.num_pages <= +maxPages
+      );
+    }
+    //* --------------- for pagination -------- *//
+    if (pages) {
+      const startIndex = (pages - 1) * limit;
+      const endIndex = pages * limit;
+
+      if (endIndex < filteredBooksData.length) {
+        response.next = {
+          page: pages + 1,
+          limit,
+        };
+      }
+      if (startIndex > 0) {
+        response.previous = {
+          page: pages - 1,
+          limit,
+        };
+      }
+      response.response = filteredBooksData.slice(startIndex, endIndex);
+      res.paginatedResults = response;
+      next();
+    } else {
+      // for sending response without pagination
+      response.response = filteredBooksData;
+      res.paginatedResults = response;
+      next();
+    }
+  };
+}
 
 // Start defining your routes here
+
+// default route
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  res.send(
+    'Welcome to the books api. For detailed documentation please visit:  '
+  );
+});
+
+// this will list all routes
+app.get('/endpoints', (req, res) => {
+  res.send(listEndPoints(app));
+});
+
+// route for all books or books based on a specific search criteria
+app.get('/books', filteredAndPaginatedResults(booksData), (req, res) => {
+  res.json({
+    response: res.paginatedResults,
+    success: true,
+  });
+});
+
+// route for the books having  the specified rating
+app.get('/books/rating/:rating', (req, res) => {
```

Comment:
It is recommended when using path params, to do an if statement so if the data does not arrive or there is a problem with it, the user or frontend developers can get a message and also the 404 status to know what is going on!
Also, when using path params, the correct method to use is the find() method, since you are expecting to get only one single element. On the other hand, the filter method is used for the query params because you expect more than one element that passes the test.